### DISCUSSION TITLE
How can I obtain the computation graph or MLIR file generated by JIT optimization?

### DISCUSSION THREAD

**Discussion by knightXun** [jax-ml/jax#21382] [Q&A] [answered]
For code like : ```output2, _ = jax.jit(model.apply)(variables, input)``` , I want to obtain the XLA computation graph or mlir code to analyse the middle optimization. How can I do it?

**Comment by jakevdp** [comment]
You can output the compiled HLO using the [Ahead of time lowering and compilation](https://jax.readthedocs.io/en/latest/aot.html) tools. It looks something like this:
```python
print(jax.jit(model.apply).lower(variables, input).compile().as_text()
```

**Reply by knightXun** [reply]
how about the uncompiled HLO?

**Reply by jakevdp** [reply, marked as answer]
Remove `.compile()` to print the uncompiled HLO text

**Reply by knightXun** [reply]
How can I print out the xla optimization steps it has taken and specify that it only performs specific xla optimization steps?

**Reply by jakevdp** [reply]
I don't know of any way to access that information from the JAX python layer.

**Reply by knightXun** [reply]
Does JAX directly send the source code to XLA for translation, or is the code already translated into HLO and then handed over to XLA for optimization? If it is translated into HLO first, how can I obtain this HLO?

**Reply by jakevdp** [reply]
JAX translates the Python to HLO before sending it to XLA for optimization. You can obtain this HLO the way I mentioned above, removing `.compile()` from the original snippet:
```python
print(jax.jit(model.apply).lower(variables, input).as_text()
```

**Reply by knightXun** [reply]
thanks!

**Reply by knightXun** [reply]
Is the 'jax/jax/interpreters' directory the implementation of code that translates Python into HLO?  Or anyothers
